File: SOD_ABC_cost.py
import os
import sys
import numpy as np
import pandas as pd
import io
import random
import grass.script as gs


def read_points(name):
    data = gs.read_command(
        "v.out.ascii",
        flags="c",
        input=name,
        columns="*",
        format="point",
        separator="comma",
    ).strip()
    return pd.read_csv(io.StringIO(data))


def minmax_scale(df, column):
    if df[column].max() - df[column].min() > 0:
        df[f"{column}_norm"] = (df[column] - df[column].min()) / (
            df[column].max() - df[column].min()
        )
    else:
        df[f"{column}_norm"] = 0
    return f"{column}_norm"


def prepare_treatments(df, treatment_map):
    treatments_string = df.to_csv(None, header=False, index=False)
    gs.write_command(
        "r.in.xyz",
        input="-",
        stdin=treatments_string,
        output=treatment_map,
        method="n",
        separator="comma",
        type="CELL",
        overwrite=True,
        quiet=True,
    )


def prepare_treatments_buffer(
    df, treatment_map, distance, distance_column, cost_raster
):
    treatments_string = df.to_csv(None, header=True, index=False)
    header = treatments_string.splitlines()[0].split(",")

    region = gs.region()
    gs.write_command(
        "v.in.ascii",
        input="-",
        stdin=treatments_string,
        output=f"{treatment_map}_points",
        separator="comma",
        skip=1,
        overwrite=True,
        quiet=True,
    )
    if distance_column:
        table_column_index = header.index(distance_column) + 1
        column = (
            gs.read_command("db.columns", table=f"{treatment_map}_points")
            .strip()
            .splitlines()[table_column_index]
        )
        gs.run_command(
            "v.buffer",
            input=f"{treatment_map}_points",
            output=f"{treatment_map}_buffers",
            column=column,
            layer=1,
            overwrite=True,
            quiet=True,
        )
    else:
        gs.run_command(
            "v.buffer",
            input=f"{treatment_map}_points",
            output=f"{treatment_map}_buffers",
            distance=distance,
            overwrite=True,
            quiet=True,
        )
    gs.use_temp_region()
    gs.run_command("g.region", res=region["nsres"] / 10, flags="a")
    gs.run_command(
        "v.to.rast",
        input=f"{treatment_map}_buffers",
        output=f"{treatment_map}_buffers",
        use="val",
        overwrite=True,
        quiet=True,
    )
    gs.del_temp_region()
    gs.run_command(
        "r.resamp.stats",
        input=f"{treatment_map}_buffers",
        output=f"{treatment_map}_count",
        method="count",
        overwrite=True,
        quiet=True,
    )
    info = gs.raster_info(f"{treatment_map}_count")
    gs.mapcalc(
        f"{treatment_map} = {treatment_map}_count / {info['max']}",
        quiet=True,
        overwrite=True,
    )
    gs.mapcalc(
        f"{treatment_map}_cost = {treatment_map} * {cost_raster}",
        quiet=True,
        overwrite=True,
    )
    actual_cost = float(
        gs.parse_command("r.univar", map=f"{treatment_map}_cost", flags="g")["sum"]
    )
    return actual_cost


def pops(treatments, average, weather_file, nprocs):
    gs.run_command(
        "r.pops.spread",
        host="lemma_den100m",
        total_plants="lemma_max100m",
        infected="cum_inf_2019",
        start_date="2019-01-01",
        end_date="2023-12-31",
        step_unit="week",
        treatments=treatments,
        treatment_date="2019-03-01",
        treatment_length=0,
        treatment_application="all_infected_in_cell",
        reproductive_rate=3.31,
        natural_distance=79.04,
        natural_direction_strength=0.47,
        natural_direction="N",
        natural_dispersal_kernel="exponential",
        seasonality=[1, 12],
        weather_coefficient_file=weather_file,
        average=average,
        output_frequency="every_n_steps",
        output_frequency_n="400",
        random_seed=1,
        runs=nprocs,
        nprocs=nprocs,
        quiet=True,
        overwrite=True,
    )
    infected_area = float(gs.raster_info(average)["description"].strip('"').split()[-1])
    return infected_area


def select_cats_by_cost(weights, costs, budget):
    tmp_budget = 0
    selected = []
    tries = 0
    weight_dict = weights.copy()
    cats = list(weight_dict.keys())
    weights = list(weight_dict.values())
    while True:
        cat = random.choices(cats, weights)[0]
        if tmp_budget + costs[cat] <= budget:
            tmp_budget += costs[cat]
            selected.append(cat)
            del weights[cats.index(cat)]
            cats.remove(cat)
        else:
            tries += 1
            if tries > 50:
                break
    return selected


def generation(
    df,
    treatment_map,
    average_map,
    baseline_evaluated,
    threshold,
    min_particles,
    weights,
    costs,
    budget,
    percentile,
    weather_file,
    buffer_distance,
    buffer_distance_column,
    cost_raster,
    nprocs,
):
    tested = 0
    new_weights = {key: 0 for key in weights}
    evaluated_list = []
    best_candidate_df = None
    minim_evaluated = baseline_evaluated
    actual_cost = []

    while len(evaluated_list) < min_particles:
        candidate_cats = select_cats_by_cost(weights, costs, budget)
        candidate_df = df.loc[df.cat.isin(candidate_cats)]
        if buffer_distance or buffer_distance_column:
            actual_cost.append(
                prepare_treatments_buffer(
                    candidate_df,
                    treatment_map,
                    buffer_distance,
                    buffer_distance_column,
                    cost_raster,
                )
            )
        else:
            prepare_treatments(candidate_df, treatment_map)
        evaluated = pops(treatment_map, average_map, weather_file, nprocs)
        tested += 1
        if evaluated / baseline_evaluated < threshold:
            for cat in candidate_cats:
                new_weights[cat] += 1
            evaluated_list.append(evaluated)
            # save minimum solution
            if evaluated < minim_evaluated:
                minim_evaluated = evaluated
                best_candidate_df = candidate_df

        acceptance_rate = len(evaluated_list) / tested

        if tested % 10 == 0:
            minim = median = None
            if evaluated_list:
                minim = np.min(evaluated_list) / baseline_evaluated
                median = np.median(evaluated_list) / baseline_evaluated
            print(
                f"Rate: {acceptance_rate}, particles left: {min_particles - len(evaluated_list)}, min evaluated: {minim}, median evaluated: {median}"
            )
    perc = np.percentile(evaluated_list, percentile)
    return (
        new_weights,
        acceptance_rate,
        perc / baseline_evaluated,
        (minim_evaluated, best_candidate_df),
        actual_cost,
    )


def estimate_initial_threshold(
    df,
    treatment_map,
    average_map,
    baseline_evaluated,
    runs,
    costs,
    budget,
    weather_file,
    buffer_distance,
    buffer_distance_column,
    cost_raster,
    nprocs,
):
    weights = {key: 1 for key in costs}
    evaluated = []
    for run in range(runs):
        candidate_cats = select_cats_by_cost(weights, costs, budget)
        candidate_df = df.loc[df.cat.isin(candidate_cats)]
        if buffer_distance or buffer_distance_column:
            prepare_treatments_buffer(
                candidate_df,
                treatment_map,
                buffer_distance,
                buffer_distance_column,
                cost_raster,
            )
        else:
            prepare_treatments(candidate_df, treatment_map)
        evaluated.append(pops(treatment_map, average_map, weather_file, nprocs))
    return np.percentile(evaluated, 10) / baseline_evaluated


def filter_particles(df, weights, iteration, percentile):
    new_df = df.copy()
    new_df["weights"] = new_df["cat"].map(weights)
    # filter unsuccessfull ones
    new_df = new_df.loc[new_df.weights > 1]
    weights_percentile = np.percentile(new_df.weights, percentile)
    # filter unlikely ones
    new_df = new_df.loc[new_df.weights >= weights_percentile]
    return new_df


def best_guess(
    df,
    weight_column,
    cost_column,
    budget,
    treatment_map,
    average_map,
    buffer_distance,
    buffer_distance_column,
    cost_raster,
    weather_file,
    nprocs,
):
    sorted_df = df.sort_values(weight_column, ascending=False)
    candidate_df = sorted_df[sorted_df[cost_column].cumsum() <= budget]
    if buffer_distance or buffer_distance_column:
        prepare_treatments_buffer(
            candidate_df,
            treatment_map,
            buffer_distance,
            buffer_distance_column,
            cost_raster,
        )
    else:
        prepare_treatments(candidate_df, treatment_map)
    candidate_df.to_csv(os.path.join(output, "best_guess.csv"))
    return pops(treatment_map, average_map, weather_file, nprocs)


def main(
    infected,
    potential_column,
    cost_column,
    cost_raster,
    buffer_distance,
    buffer_distance_column,
    budget,
    min_particles,
    filter_percentile,
    threshold_percentile,
    output,
    nprocs,
):
    treatment_map = gs.append_node_pid("treatments")
    average_map = gs.append_node_pid("average")
    weather_file = gs.tempfile(create=False)
    gs.run_command(
        "g.list", mapset="weather", flags="m", type="raster", output=weather_file
    )

    df = read_points(infected)
    potential_norm_column = minmax_scale(df, potential_column)
    cost_norm_column = minmax_scale(df, cost_column)
    df["prior_weight"] = (df[potential_norm_column] + 1 - df[cost_norm_column]) / 2

    initial_evaluated = pops(None, average_map, weather_file, nprocs)
    print(f"Initial evaluated: {initial_evaluated}")

    best_guess_evaluated = best_guess(
        df,
        "prior_weight",
        cost_column,
        budget,
        treatment_map,
        average_map,
        buffer_distance,
        buffer_distance_column,
        cost_raster,
        weather_file,
        nprocs,
    )
    print(f"Best guess: {best_guess_evaluated}")
    weights = dict(zip(df.cat, df.prior_weight))
    costs = dict(zip(df.cat, df[cost_column]))

    thresholds = []
    thresholds.append(
        estimate_initial_threshold(
            df,
            treatment_map,
            average_map,
            initial_evaluated,
            10,
            costs,
            budget,
            weather_file,
            buffer_distance,
            buffer_distance_column,
            cost_raster,
            nprocs,
        )
    )
    acceptance_rates = []
    actual_cost_mean = []
    filtered_dfs = []
    tmpdf = df.copy()
    iteration = 0
    while True:
        print(f"Iteration {iteration}: threshold {thresholds[-1]}")
        weights, acceptance_rate, perc_evaluated, best, actual_cost = generation(
            df,
            treatment_map,
            average_map,
            initial_evaluated,
            thresholds[-1],
            min_particles,
            weights,
            costs,
            budget,
            threshold_percentile,
            weather_file,
            buffer_distance,
            buffer_distance_column,
            cost_raster,
            nprocs,
        )
        acceptance_rates.append(acceptance_rate)
        thresholds.append(perc_evaluated)
        actual_cost_mean.append(
            sum(actual_cost) / len(actual_cost) if len(actual_cost) else None
        )
        filtered_df = filter_particles(tmpdf, weights, iteration, filter_percentile)
        if filtered_df[cost_column].sum() >= budget:
            print(f"Filtered {len(tmpdf) - len(filtered_df)} pixels from {len(tmpdf)}")
            tmpdf = filtered_df
            weights = dict(zip(tmpdf.cat, tmpdf.weights))
            filtered_dfs.append(filtered_df)
        else:
            break
        iteration += 1

    print(f"Best evaluated: {best[0]}")
    best[-1].to_csv(os.path.join(output, "best.csv"))
    tmpdf.to_csv(os.path.join(output, "last_filtered.csv"))
    print(f"Acceptance rate: {acceptance_rates}")
    print(f"Thresholds: {thresholds}")
    print(f"Actual cost: {actual_cost_mean}")


if __name__ == "__main__":
    budget = float(sys.argv[1])
    cost = sys.argv[2]
    min_particles = int(sys.argv[3])
    filter_percentile = float(sys.argv[4])
    threshold_percentile = float(sys.argv[5])
    try:
        buffer_distance = float(sys.argv[6])
        buffer_distance = buffer_distance if buffer_distance > 0 else None
        buffer_distance_column = None
    except ValueError:
        buffer_distance = None
        buffer_distance_column = sys.argv[6]
    output = sys.argv[7]
    nprocs = int(sys.argv[8])
    infected = "infected_2019"
    potential = "potential"
    cost_raster = "walk_cost"
    main(
        infected,
        potential,
        cost,
        cost_raster,
        buffer_distance,
        buffer_distance_column,
        budget,
        min_particles,
        filter_percentile,
        threshold_percentile,
        output,
        nprocs,
    )

File: SOD_ABC_cost.R
read_points <- function(name) {
  return(read_VECT(name, layer = "1"))
  # return (as.data.frame(data, geom="XY"))
}

get_norm_column_name <- function(column) {
  return(paste0(column, "_norm"))
}

minmax_scale <- function(df, column) {
  new_column <- get_norm_column_name(column)
  max_val <- max(df[[column]])
  min_val <- min(df[[column]])

  if (max_val - min_val > 0) {
    df[[new_column]] <- scales::rescale(df[[column]][[1]], to = c(0, 1))
  } else {
    df[[new_column]] <- 0
  }
  return(df)
}

prepare_treatments <- function(df, treatment_map, cost_raster) {
  r <- terra::rast(terra::ext(cost_raster), resolution = terra::res(cost_raster))
  terra::crs(r) <- terra::crs(cost_raster)
  write_RAST(terra::rasterize(df, r), treatment_map, flags = "overwrite")
}

prepare_treatments_buffer <- function(df, treatment_map, cost_raster, distance, distance_column) {
  rasterization_factor <- 10
  if (is.null(distance_column)) {
    buffers <- terra::buffer(df, width = distance)
  } else {
    buffers <- terra::buffer(df, width = df[[distance_column]][[1]])
  }
  r <- terra::rast(terra::ext(cost_raster), resolution = terra::res(cost_raster))
  terra::crs(r) <- terra::crs(cost_raster)
  high_res_raster <- terra::disagg(r, fact = rasterization_factor)
  buffers_raster <- terra::rasterize(buffers, high_res_raster)
  count_raster <- terra::aggregate(buffers_raster,
    fact = rasterization_factor,
    fun = function(x) {
      length(x[!is.na(x)])
    }
  )
  treatment_raster <- count_raster / terra::global(count_raster, "max")[[1]]
  treatment_cost_raster <- treatment_raster * cost_raster
  write_RAST(treatment_raster, treatment_map, flags = "overwrite")
  actual_cost <- terra::global(treatment_cost_raster, "sum", na.rm = T)[[1]]
  return(actual_cost)
}

pops <- function(treatments, average, weather_file, nprocs) {
  execGRASS(
    "r.pops.spread",
    model_type = "SI",
    host = "lemma_den100m",
    total_plants = "lemma_max100m",
    infected = "cum_inf_2019",
    start_date = "2019-01-01",
    end_date = "2023-12-31",
    step_unit = "week",
    step_num_units = 1,
    treatments = treatments,
    treatment_date = "2019-03-01",
    treatment_length = 0,
    treatment_application = "all_infected_in_cell",
    reproductive_rate = 1.6,
    natural_distance = 25,
    anthropogenic_distance = 4000,
    natural_direction_strength = 1,
    anthropogenic_direction_strength = 1,
    natural_direction = "N",
    anthropogenic_direction = "N",
    natural_dispersal_kernel = "exponential",
    anthropogenic_dispersal_kernel = "cauchy",
    percent_natural_dispersal = 0.1,
    seasonality = "1,12",
    weather_coefficient_file = weather_file,
    average = average,
    output_frequency = "every_n_steps",
    output_frequency_n = 400,
    random_seed = 1,
    runs = nprocs,
    nprocs = nprocs,
    flags = c("overwrite", "quiet")
  )
  infected_area <- execGRASS("r.info", map = average, flags = c("g", "e"), intern = TRUE)
  infected_area <- tail(strsplit(infected_area[[26]], " ")[[1]], 1)
  infected_area <- as.numeric(gsub("\"", "", infected_area))

  return(infected_area)
}

sample_candidate <- function(df, cat_column, cost_column, weight_column, budget) {
  # to allow sample vector size 1
  sample.vec <- function(x, ...) x[sample(length(x), ...)]

  tmp_budget <- 0
  tries <- 0
  count <- 1
  candidate_cats <- c()
  cats <- df[[cat_column]][[1]]
  weights <- df[[weight_column]][[1]]
  costs <- df[[cost_column]][[1]]
  while (length(cats)) {
    sample_cat <- sample.vec(cats, 1, prob = weights)
    selected_index <- which(cats == sample_cat)
    cost <- costs[[selected_index]]
    if (tmp_budget + cost <= budget) {
      tmp_budget <- tmp_budget + cost
      candidate_cats[count] <- sample_cat
      count <- count + 1
      cats <- cats[-selected_index]
      weights <- weights[-selected_index]
      costs <- costs[-selected_index]
    } else {
      tries <- tries + 1
      if (tries > 50) {
        break
      }
    }
  }
  return(df[df[[cat_column]][[1]] %in% candidate_cats, ])
}

generation <- function(df,
                       treatment_map,
                       average_map,
                       baseline_evaluated,
                       threshold,
                       min_particles,
                       cat_column,
                       cost_column,
                       weight_column,
                       budget,
                       percentile,
                       weather_file,
                       buffer_distance,
                       buffer_distance_column,
                       cost_raster,
                       nprocs) {
  evaluated_list <- c()
  actual_cost <- c()
  evaluated_index <- 1
  tested <- 0
  cats <- df[[cat_column]][[1]]
  new_weights <- setNames(as.list(rep(0, length(cats))), cats)
  minim_evaluated <- baseline_evaluated
  best_candidate_df <- NULL
  while (length(evaluated_list) < min_particles) {
    candidate_df <- sample_candidate(df, cat_column, cost_column, weight_column, budget)

    if (!is.null(buffer_distance) | !is.null(buffer_distance_column)) {
      cost <- prepare_treatments_buffer(
        candidate_df,
        treatment_map,
        cost_raster,
        buffer_distance,
        buffer_distance_column
      )
      actual_cost[evaluated_index] <- cost
      evaluated_index <- evaluated_index + 1
    } else {
      prepare_treatments(candidate_df, treatment_map, cost_raster)
    }

    evaluated <- pops(treatment_map, average_map, weather_file, nprocs)
    tested <- tested + 1
    print(evaluated / baseline_evaluated)
    if (evaluated / baseline_evaluated < threshold) {
      for (cat in as.character(candidate_df[[cat_column]][[1]])) {
        new_weights[[cat]] <- new_weights[[cat]] + 1
      }
      evaluated_list[evaluated_index] <- evaluated
      evaluated_index <- evaluated_index + 1
      if (evaluated < minim_evaluated) {
        minim_evaluated <- evaluated
        best_candidate_df <- candidate_df
      }
    }
    acceptance_rate <- length(evaluated_list) / tested
    if (tested %% 10 == 0) {
      minim <- NULL
      median <- NULL
      if (length(evaluated_list) != 0) {
        minim <- min(evaluated_list) / baseline_evaluated
        median <- median(evaluated_list) / baseline_evaluated
        print(paste(
          "Rate:", acceptance_rate, "particles left:", min_particles - length(evaluated_list),
          "min evaluated:", minim, "median evaluated:", median
        ))
      }
    }
  }
  perc <- quantile(evaluated_list, probs = percentile / 100)
  return(list(
    new_weights, acceptance_rate, perc / baseline_evaluated,
    list(minim_evaluated, best_candidate_df), actual_cost
  ))
}


estimate_initial_threshold <- function(df,
                                       treatment_map,
                                       average_map,
                                       baseline_evaluated,
                                       cat_column,
                                       cost_column,
                                       weight_column,
                                       budget,
                                       weather_file,
                                       buffer_distance,
                                       buffer_distance_column,
                                       cost_raster,
                                       runs,
                                       nprocs) {
  evaluated_list <- c()
  for (run in seq(runs)) {
    candidate_df <- sample_candidate(df, cat_column, cost_column, weight_column, budget)

    if (!is.null(buffer_distance) | !is.null(buffer_distance_column)) {
      prepare_treatments_buffer(
        candidate_df,
        treatment_map,
        cost_raster,
        buffer_distance,
        buffer_distance_column
      )
    } else {
      prepare_treatments(candidate_df, treatment_map, cost_raster)
    }
    evaluated_list[run] <- pops(treatment_map, average_map, weather_file, nprocs)
  }
  return(quantile(evaluated_list, probs = 0.1) / baseline_evaluated)
}

best_guess <- function(df,
                       treatment_map,
                       average_map,
                       cat_column,
                       cost_column,
                       weight_column,
                       budget,
                       weather_file,
                       buffer_distance,
                       buffer_distance_column,
                       cost_raster,
                       nprocs) {
  sorted_df <- df[order(df[[weight_column]][[1]], decreasing = TRUE), ]
  candidate_df <- sorted_df[cumsum(sorted_df[[cost_column]]) <= budget, ]

  if (!is.null(buffer_distance) | !is.null(buffer_distance_column)) {
    prepare_treatments_buffer(
      candidate_df,
      treatment_map,
      cost_raster,
      buffer_distance,
      buffer_distance_column
    )
  } else {
    prepare_treatments(candidate_df, treatment_map, cost_raster)
  }
  # TODO print somewhere
  return(pops(treatment_map, average_map, weather_file, nprocs))
}

filter_particles <- function(df, cat_column, weight_column, iteration, percentile) {
  # filter unsuccessful ones
  filtered_df <- df[df[[weight_column]] > 0, ]
  weights_percentile <- quantile(filtered_df[[weight_column]][[1]],
    percentile / 100,
    names = FALSE
  )
  # filter unlikely ones
  df[df[[weight_column]] <= weights_percentile, weight_column] <- 0
  return(df)
}


library("rgrass")
library("terra")

infected <- "infected_2019"
potential_column <- "potential"
cost_column <- "pixel_cost"
cost <- "walk_cost"
cat_column <- "cat"
min_particles <- 10
budget <- 20000
filter_percentile <- 15
threshold_percentile <- 10
buffer_distance <- NULL
buffer_distance_column <- NULL
nprocs <- 10
treatment_map <- "treatment_test"
average_map <- "average_test"


weather_file <- "/tmp/weather.txt"
execGRASS("g.list",
  mapset = "weather",
  flags = c("m", "overwrite"),
  type = "raster",
  output = weather_file
)

df <- read_points(infected)

df <- minmax_scale(df, potential_column)
df <- minmax_scale(df, cost_column)
iteration <- 1
weight_column <- paste0("weight_", iteration)
df[[weight_column]] <- (df[[get_norm_column_name(potential_column)]] + 1 - df[[get_norm_column_name(cost_column)]]) / 2

cost_raster <- read_RAST(cost)

baseline_evaluated <- pops("", average_map, weather_file, nprocs)
print(paste("Baseline:", baseline_evaluated))
best_guess_evaluated <- best_guess(
  df,
  treatment_map,
  average_map,
  cat_column,
  cost_column,
  weight_column,
  budget,
  weather_file,
  buffer_distance,
  buffer_distance_column,
  cost_raster,
  nprocs
)
print(paste("Best guess:", best_guess_evaluated))

thresholds <- c()
thresholds[1] <- estimate_initial_threshold(
  df,
  treatment_map,
  average_map,
  baseline_evaluated,
  cat_column,
  cost_column,
  weight_column,
  budget,
  weather_file,
  buffer_distance,
  buffer_distance_column,
  cost_raster,
  10,
  nprocs
)


acceptance_rates <- c()
actual_cost_mean <- c()
filtered_df <- df
tmp_df <- df
while (TRUE) {
  print(paste0("Iteration ", iteration, ", threshold ", thresholds[length(thresholds)]))
  results <- generation(
    filtered_df,
    treatment_map,
    average_map,
    baseline_evaluated,
    thresholds[length(thresholds)],
    min_particles, cat_column,
    cost_column,
    weight_column,
    budget,
    threshold_percentile,
    weather_file,
    buffer_distance,
    buffer_distance_column,
    cost_raster,
    nprocs
  )
  acceptance_rates <- append(acceptance_rates, results[[2]])
  thresholds <- append(thresholds, results[[3]])
  if (length(results[[5]]) != 0) {
    actual_cost_mean <- append(actual_cost_mean, sum(results[[5]]) / len(results[[5]]))
  }
  new_weight_column <- paste0("weight_", iteration + 1)
  new_weights <- results[[1]]

  filtered_df[[new_weight_column]] <- new_weights[match(
    filtered_df[[cat_column]][[1]],
    as.integer(names(new_weights))
  )]
  tmp_df <- filter_particles(
    filtered_df,
    cat_column,
    new_weight_column,
    iteration,
    filter_percentile
  )
  before_df <- filtered_df[filtered_df[[new_weight_column]] > 0, ]
  after_df <- tmp_df[tmp_df[[new_weight_column]] > 0, ]
  if (sum(after_df[[cost_column]][[1]]) >= budget) {
    print(paste0(
      "Filtered ", length(before_df) - length(after_df),
      ", remaining: ", length(after_df)
    ))
    filtered_df <- tmp_df
    weight_column <- new_weight_column
  } else {
    break
  }
  iteration <- iteration + 1
}
print(paste0("Best evaluated: ", results[[4]][[1]]))
print(paste0("Acceptance rate: ", acceptance_rates))
print(paste0("Thresholds: ", thresholds))
print(paste0("Actual cost: ", actual_cost_mean))
